rvsol/test/RISCV.t.sol

```diff
@@ -360,6 +360,276 @@ contract RISCV_Test is CommonTest {
+        int256 temp = int256(int64(state.registers[31])) * int256(int64(state.registers[2]));
+        expect.registers[20] = uint64(uint256((temp >> 64) & ((1 << 64) - 1)));
+        expect.registers[31] = state.registers[31];
+        expect.registers[2] = state.registers[2];
+
+        bytes32 postState = riscv.step(encodedState, proof, 0);
+        assertEq(postState, outputState(expect), "unexpected post state");
+    }
+
+    function test_mulhsu_succeeds() public {
+        uint32 insn = encodeRType(0x33, 7, 2, 5, 27, 1); // mulhsu x7, x5, x27
+        (State memory state, bytes memory proof) = constructRISCVState(0, insn);
+        state.registers[5] = 0x6f050b64e4f37291;
+        state.registers[27] = 0x2b29ce113892ba69;
+        bytes memory encodedState = encodeState(state);
+
+        State memory expect;
+        expect.memRoot = state.memRoot;
+        expect.pc = state.pc + 4;
+        expect.step = state.step + 1;
+        int256 temp = int256(int64(state.registers[5])) * int256(int128(uint128(state.registers[27])));
+        expect.registers[7] = uint64(uint256((temp >> 64) & ((1 << 64) - 1)));
+        expect.registers[5] = state.registers[5];
+        expect.registers[27] = state.registers[27];
+
+        bytes32 postState = riscv.step(encodedState, proof, 0);
+        assertEq(postState, outputState(expect), "unexpected post state");
+    }
+
+    function test_mulhu_succeeds() public {
+        uint32 insn = encodeRType(0x33, 27, 3, 6, 18, 1); // mulhu x27, x6, x18
+        (State memory state, bytes memory proof) = constructRISCVState(0, insn);
+        state.registers[6] = 0x4544e15a2fb9c134;
+        state.registers[18] = 0xa11b583879ae1211;
+        bytes memory encodedState = encodeState(state);
+
+        State memory expect;
+        expect.memRoot = state.memRoot;
+        expect.pc = state.pc + 4;
+        expect.step = state.step + 1;
+        uint256 temp = uint256(state.registers[6]) * uint256(state.registers[18]);
+        expect.registers[27] = uint64(uint256((temp >> 64) & ((1 << 64) - 1)));
+        expect.registers[6] = state.registers[6];
+        expect.registers[18] = state.registers[18];
+
+        bytes32 postState = riscv.step(encodedState, proof, 0);
+        assertEq(postState, outputState(expect), "unexpected post state");
+    }
+
+    function test_div_succeeds() public {
+        uint32 insn = encodeRType(0x33, 18, 4, 4, 29, 1); // div x18, x4, x29
+        (State memory state, bytes memory proof) = constructRISCVState(0, insn);
+        state.registers[4] = 0xf0e9720e82469b35;
+        state.registers[29] = 0xf391cc717328ac0;
+        bytes memory encodedState = encodeState(state);
+
+        State memory expect;
+        expect.memRoot = state.memRoot;
+        expect.pc = state.pc + 4;
+        expect.step = state.step + 1;
+        int256 temp = int256(int64(state.registers[4])) / int256(int64(state.registers[29]));
+        expect.registers[18] = uint64(uint256(temp & ((1 << 64) - 1)));
```

Comment:
why do we need casting to 256bits for div?

Reply:
same for rem

Reply:
Casting is not mandatory for div and rem, but for mul, casting is needed to avoid overflow errors. I used the same templates to code up tests for div/rem.

Reply:
Fixed at https://github.com/ethereum-optimism/asterisc/pull/22/commits/83e43ac7ac0c0525770e8f031b47aa0e28dec1ba

Reply:
@pcw109550 sorry but can you remove casting for divu and remu as well?

Reply:
They need casting because they are unsigned.